### PR TITLE
Path/Line Integrated Diagnostic (Abstract classes + interferometer implementation)

### PR DESCRIPTION
**Pull request by pheuer**
This PR introduces a new abstract class "PathIntegratedDiagnostic" to the diagnostics module. This class abstracts the source/grid/detector geometry already used in `SyntheticProtonRadiography`. I've refactored the `SyntheticProtonRadiograph` module to be a subclass of the new class, but there are no API changes.

This PR also introuduces a subclass of `PathIntegratedDiagnostic`, `LineIntegratedDiagnostic`, which represents the case where the integration paths are straight lines. Given a `grid` object, this class can line-integrate through any quantity on the grid along any arbitrary straight line of sight. The quantity integrated is given by a "_integrand" method, which can be in turn be calculated using any combination of the quantities on the grid. Different diagnostics with non-trivial response functions (like proton radiography) can be implemented by just replacing this `_integrand` in subclasses.

This module could has many applications (x-ray radiography, linear-regime charged particle radiography, etc.), but for a start I've implemented a very basic interferometry function. Given a 1 mm radius sphere of constant density 3e19 cm^-3, the following is the predicted interferogram with a 263 nm probe beam

interferogram

The documentation still needs some work (I'll write one notebook explaining the interferometry subclass and another showing how to create and use custom subclasses). But I'm ready for comments on the basic structure of the `LineIntegratedDiagnostic` class!

### CODE REVIEW

**Comment by pheuer** [comment]
This PR has a couple errors in the checks that I don't understand: 

```
./plasmapy/diagnostics/line_integrated_diagnostic.py:62:1: RST301 Unexpected indentation.
    def _log(self, msg):
^
./plasmapy/diagnostics/line_integrated_diagnostic.py:214:1: RST215 Inline interpreted text or phrase reference start-string without end-string.
        return (xax * u.m, yax * u.m, *integral)
^
./plasmapy/diagnostics/line_integrated_diagnostic.py:272:1: RST215 Inline interpreted text or phrase reference start-string without end-string.
        # Continue with the rest of the parent class init
^
2     RST215 Inline interpreted text or phrase reference start-string without end-string.
1     RST301 Unexpected indentation.
```

**Comment by StanczakDominik** [comment]
Hey, sorry, that's flake8 being overzealous after #1062 ; ignore the flake8 for now, please :( I'll get to it when I'm less swamped :(

**Comment by StanczakDominik** [comment]
All right, things have calmed down, I'll give this a thorough review tomorrow. Still need to recharge my batteries a little :(

**Comment by pheuer** [comment]
Sounds good - no rush though! 

Note that I've changed the naming a bit: 

"PathIntegratedDiagnostic" -> Encapsulates the source/detector/grid geometry setup.

"LineIntegratedDiagnostic" -> Subclass for specifically the linear-integration regime, include `_line_integral()`.

File: docs/notebooks/diagnostics/line_integrated_diagnostic.ipynb [hunk 1]
```diff
@@ -0,0 +1,341 @@
+{
```
Comment (StanczakDominik) [inline docs/notebooks/diagnostics/line_integrated_diagnostic.ipynb:1]: <p>Just starting out by reading the notebook for now, but - I suppose the use of CartesianGrid here is more along the lines of "it's set up to work with CartesianGrid right now", rather than "It will never take a non-uniform grid"? Perhaps we could reword it to:</p><p> </p><p>"by providing a Grid object, representing the plasma to be measured (currently only CartesianGrid is supported)"</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/PlasmaPy/PlasmaPy/pull/1123/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

Comment (StanczakDominik) [inline docs/notebooks/diagnostics/line_integrated_diagnostic.ipynb:1]: <p>My first impression when reading this whole heading was "Oh, I have to create a custom class to use this? I wonder if we could not make this easier to use somehow" - and then the last sentence up-ends my understanding! Perhaps we could reword the title to something like "Example: custom Interferometer class", and in the beginning of this part say something about how this is mostly a didactic example so people can wrap their head around the pre-made <code>Interferometer</code> and the whole LineIntegratedDiagnostic framework easier?</p><p> </p><p>It would also be helpful to point people to a place in the docs where they can find e.g. a schema for the arguments of <code>evaluate</code>.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/PlasmaPy/PlasmaPy/pull/1123/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

Comment (StanczakDominik) [inline docs/notebooks/diagnostics/line_integrated_diagnostic.ipynb:1]: <p>Could do with a title noting that the plot is a projection to a plane perpendicular to the line between the source and detector.</p><p> </p><p>Eh, you know what we should do, actually? Add pyvista as at least a docs dependency. It's 2021, let's do 3D plotting already. I volunteer to do that and modernise our plots after I wrap up with my thesis.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/PlasmaPy/PlasmaPy/pull/1123/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

Comment (StanczakDominik) [inline docs/notebooks/diagnostics/line_integrated_diagnostic.ipynb:1]: <p>It would be very nice to have a repeat of the same thing with a higher resolution detector plate, and an explanation of how the pixelated distribution is the result of the chosen resolution.</p><p> </p><p>By the way, if you end the last line with </p><p><code>cb.ax.set_ylabel("Phase Shift / $\pi$", fontsize=16); #note the semicolon</code></p><p> </p><p>It will suppress the output of <code>Text(0, 0.5, 'Phase Shift / $\\pi$')</code>  as the last object returned from the notebook cell.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/PlasmaPy/PlasmaPy/pull/1123/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='17'/>

Comment (StanczakDominik) [inline docs/notebooks/diagnostics/line_integrated_diagnostic.ipynb:1]: <p>As a newbie to interferometry, and speaking for all newbies to interferometry to come after me - some more explanation of the unwrapping could be handy! :D I have a hunch about how it works, and I expect diving through the code will confirm it, but just noting a note of confusion down.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/PlasmaPy/PlasmaPy/pull/1123/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='18'/>

File: plasmapy/diagnostics/tests/test_path_integrated_diagnostic.py [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+    size = np.array([[-2, 2], [-2, 2]]) * u.mm
+    bins = [350, 350]
+
+    hax, vax, phase = obj.evaluate(
+        1.14e15 * u.Hz,
+        size=size,
+        bins=bins,
+        num=100,
+        unwrapped=False,
+    )
```
Comment (StanczakDominik) [inline plasmapy/diagnostics/tests/test_path_integrated_diagnostic.py:162]: Just checking that this evaluates?

Would it help if we had mechanisms in place to verify that our tests keep returning the same image as they did previously? I'm asking because I've been putting #779 off for a while (started dabbling it a little while ~~frustrated with~~ working on https://github.com/PlasmaPy/PlasmaPy/pull/1059 ), and I could get back to it if it made this kind of test easier.

File: plasmapy/diagnostics/path_integrated_diagnostic.py [hunk 3]
```diff
@@ -0,0 +1,412 @@
...
+            Number of bins in each direction of the detector region. The
+            default is [50,50].
+
+        collimated : Boolean, optional
+            If True, the source will be assumed to be collimated. If False,
+            a point source will be used. The default is True (collimated).
+
+        num : int, optional
+            Number of integration points along the line (within the grid region).
+            The default is 100.
```
Comment (StanczakDominik) [inline plasmapy/diagnostics/path_integrated_diagnostic.py:185]: Seems like a potentially fragile default...? I imagine if someone has steep gradients, it might cause incorrect results.

**Review by StanczakDominik — COMMENTED** [review]
Looks good on the architecture side! A couple questions above about the notebook, and a few here.

File: docs/notebooks/diagnostics/line_integrated_diagnostic.ipynb (hunk 1 above)
Comment (pheuer) [reply docs/notebooks/diagnostics/line_integrated_diagnostic.ipynb:1]: Actually, this should already work for non-uniform grids! CartesianGrids are just so much easier to set up. But maybe I should add an example with a non-uniform grid too...

Comment (pheuer) [reply docs/notebooks/diagnostics/line_integrated_diagnostic.ipynb:1]: Ah, good point! I've added the following to the beginning of that section to hopefully head off the confusion: 

```
A [LineIntegratedDiagnostic] synthetic diagnostic for interferometry ([Interferometer]) is already implemented in PlasmaPy. However, in this notebook we will re-create this class in order to demonstrate how a custom subclass of [LineIntegratedDiagnostic] can be used to create a new synthetic diagnostic.
```

There should probably be a separate (short) notebook explaining the actual `Interferometer` class? That would hopefully also help dispel the confusion and provide a more natural place to document the `evaluate` method.

Comment (pheuer) [reply docs/notebooks/diagnostics/line_integrated_diagnostic.ipynb:1]: I'm not sure I follow this comment. I agree making that more clear is a good, but is a plot title the best way? Rather than putting it in the text?

Comment (pheuer) [reply docs/notebooks/diagnostics/line_integrated_diagnostic.ipynb:1]: Thanks for the semicolon tip! 

Unfortunately this is already pretty high resolution: I'll double it but it doesn't look THAT much better. The restriction is the size of the array that we need to evaluate the integrand at: for a [h#, v#] detector plane and num=[n#] integration points that's an [h#, v#, n#] array, so it can become slow if the resolution gets high.

Comment (StanczakDominik) [reply docs/notebooks/diagnostics/line_integrated_diagnostic.ipynb:1]: Ah, in that case, let's just reword it! :)

Comment (StanczakDominik) [reply docs/notebooks/diagnostics/line_integrated_diagnostic.ipynb:1]: It's probably fine to keep the example in this notebook. If/once it grows out, we can add another one.

Comment (StanczakDominik) [reply docs/notebooks/diagnostics/line_integrated_diagnostic.ipynb:1]: Ah, sorry. The plot title was just the first thing that came to mind. Some way of communicating "this situation is 3D, but the reason we're seeing a 2D plot is that we're projecting to the XZ (IIRC) plane" would be great :)

Comment (pheuer) [reply docs/notebooks/diagnostics/line_integrated_diagnostic.ipynb:1]: So, basically you just remove discontinuities of 2pi using `np.unwrap`. 

I think this belongs in the Interferometer notebook, so I'm starting that now and I'll manually unwrap the signal there as an example.

Comment (StanczakDominik) [reply docs/notebooks/diagnostics/line_integrated_diagnostic.ipynb:1]: In that case we could go the other way - start with a smaller resolution and compare with the current one?

File: plasmapy/diagnostics/path_integrated_diagnostic.py (hunk 3 above)
Comment (pheuer) [reply plasmapy/diagnostics/path_integrated_diagnostic.py:185]: Maybe it makes sense to just make this a required input? 100 could be overkill for some grids or insufficient for others. Or it could default to `npmax(grid.shape)` or something?

File: docs/notebooks/diagnostics/line_integrated_diagnostic.ipynb (hunk 1 above)
Comment (pheuer) [reply docs/notebooks/diagnostics/line_integrated_diagnostic.ipynb:1]: Ok, I'll keep it like that for now. Eventually there will be a bunch of these LineIntegratedDiagnostics, so then it may make sense for them each have a notebook?

File: plasmapy/diagnostics/tests/test_path_integrated_diagnostic.py [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+
+    assert np.allclose(line, theory, atol=5e-4)
+
+
+def test_constant_box(grid):
+    x, y, z = grid.grid.T
+    t1 = np.where(np.abs(x) < 1 * u.mm, 1, 0)
+    t2 = np.where(np.abs(z) < 1 * u.mm, 1, 0)
+    field = np.where(t1 * t2 != 0, 1, 0) * u.kg / u.m ** 3
+    grid.add_quantities(rho=field)
```
[the remark this answers is not in the document]
Comment (pheuer) [reply plasmapy/diagnostics/tests/test_path_integrated_diagnostic.py:133]: What method? Seems to be working to me!

File: plasmapy/diagnostics/tests/test_path_integrated_diagnostic.py (hunk 2 above)
Comment (pheuer) [reply plasmapy/diagnostics/tests/test_path_integrated_diagnostic.py:183]: Yes I would use that all the time! Ideally it would include some sort of `np.isclose` rtol keyword or something. As is I've been writing custom checks based on the geometry of each image and I just haven't thought of one for this one yet.

File: plasmapy/diagnostics/tests/test_path_integrated_diagnostic.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (StanczakDominik) [reply plasmapy/diagnostics/tests/test_path_integrated_diagnostic.py:133]: I mean `grid.add_quantities(rho=field)` if rho already exists! Glad it's working, though!

File: plasmapy/diagnostics/tests/test_path_integrated_diagnostic.py (hunk 2 above)
Comment (StanczakDominik) [reply plasmapy/diagnostics/tests/test_path_integrated_diagnostic.py]: All right, I'll try to work on that this weekend, then :)

File: plasmapy/diagnostics/tests/test_path_integrated_diagnostic.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (pheuer) [reply plasmapy/diagnostics/tests/test_path_integrated_diagnostic.py:133]: Oh, yes I think that should just overwrite whatever was there first!

**Comment by StanczakDominik** [comment]
Whoa. Take a look at the RTD error. The notebook kernel is dying. Looks like it's high time we pre-executed the longer-running simulation-based notebooks...

**Comment by StanczakDominik** [comment]
As a stop gap solution, I commit to rerunning them on a monthly basis to check if anything's broken. I imagine I'll be pissed off enough to find an automatic solution within 3 months or so. :D

**Comment by pheuer** [comment]
Thanks @StanczakDominik. Do you or @rocco8773 know why CI Documentation is raising the following error still? 

```
Warning, treated as error:
/home/runner/work/PlasmaPy/PlasmaPy/docs/ad/diagnostics/path_integrated_diagnostic.rst:9:An option to automodapi is either unknown or has an invalid value: 'no-heading'
ERROR: InvocationError for command /home/runner/work/PlasmaPy/PlasmaPy/.tox/build_docs/bin/sphinx-build docs docs/_build/html -W -b html (exited with code 2)
```

The offending line 
```
.. automodapi:: plasmapy.diagnostics.path_integrated_diagnostic
   :no-heading:
```

It looks like `:no-heading:` is in all the files, so I don't know why it's causing a problem here?

**Comment by StanczakDominik** [comment]
I don't know why it gave a warning in f714472, but I'm not seeing any of those on master, so I think @rocco8773 dropped it with #1105. I got rid of it; let's see how it renders.

File: plasmapy/diagnostics/path_integrated_diagnostic.py [hunk 5]
```diff
@@ -0,0 +1,627 @@
...
+        location and the size and bins keywords.
+
+        2) For each cell of detector grid, create an array of points of
+        separation ds from there to the source point (or, when collimated=True,
+        the source plane).
+
+        3) Evaluate the integrand function at each point.
+
+        4) Integrate along the line-integrated dimension to obtain the
+        line-integrated quantity in the detector plane.
```
Comment (rocco8773) [inline plasmapy/diagnostics/path_integrated_diagnostic.py:498]: This is probably causing the new errors since the line continuation is suppose to indent.

File: plasmapy/diagnostics/path_integrated_diagnostic.py [hunk 6]
```diff
@@ -0,0 +1,627 @@
...
+        bins=[200, 200],
+        collimated=True,
+        unwrapped=True,
+    ):
+        r"""
+        Creates an interferogram by calculating the line integral through
+        the electron number density :math:`n_e` provided on the grid. The phase shift
+        is
+
+        .. math:: \Delta \phi = -\frac{\omega_{probe}}{2 c n_c} \int n_e dl
```
Comment (rocco8773) [inline plasmapy/diagnostics/path_integrated_diagnostic.py:556]: If you don't want to make this  raw string, then you'll have to put an escape character on math functions; that is, 

```rst
 .. math:: \\Delta \\phi = -\\frac{\\omega_{probe}}{2 c n_c} \\int n_e dl
```

File: plasmapy/diagnostics/path_integrated_diagnostic.py [hunk 7]
```diff
@@ -0,0 +1,627 @@
...
+
+    def evaluate(
+        self,
+        num,
+        size=None,
+        bins=[200, 200],
+        collimated=True,
+    ):
+
+        r"""
```
Comment (rocco8773) [inline plasmapy/diagnostics/path_integrated_diagnostic.py:484]:
```suggestion
    ):
        r"""
```

**Comment by pheuer** [comment]
@StanczakDominik This error is appearing intermittently 
```
FAILED plasmapy/formulary/tests/test_Chandrasekhar_G.py::test_Chandrasekhar_with_hypothesis - assert 0.25 <= 0.21399915915288345
```

**Comment by StanczakDominik** [comment]
Welp that's on me. Feel free to ignore it and I will deal with it tomorrow

File: plasmapy/diagnostics/path_integrated_diagnostic.py (hunk 6 above)
Comment (pheuer) [reply plasmapy/diagnostics/path_integrated_diagnostic.py]: Oh ok, done

File: docs/notebooks/diagnostics/line_integrated_diagnostic.ipynb [hunk 8]
```diff
@@ -0,0 +1,1032 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "weekly-howard",
+   "metadata": {},
+   "source": [
+    "[LineIntegratedDiagnostic]: ../../api/plasmapy.diagnostics.path_integrated_diagnostic.LineIntegratedDiagnostic.rst#plasmapy.diagnostics.path_integrated_diagnostic.LineIntegratedDiagnostic\n",
+    "\n",
+    "[Grid]: ../../api/plasmapy.plasma.grids.rst#plasmapy.plasma.grids\n",
```
Comment (rocco8773) [inline docs/notebooks/diagnostics/line_integrated_diagnostic.ipynb:71]:
```suggestion
    "[Grid]: ../../api_static/plasmapy.plasma.grids.rst\n",
```

This will get rid of the next doc build error.

File: docs/notebooks/diagnostics/line_integrated_diagnostic.ipynb [hunk 9]
```diff
@@ -0,0 +1,1032 @@
...
+   "source": [
+    "## Interferometer"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "dfaa16bd",
+   "metadata": {},
+   "source": [
+    "[Interferometer]: ../../api/plasmapy.diagnostics.path_integrated_diagnostic.interferometry.Interferometer.rst#plasmapy.diagnostics.path_integrated_diagnostic.interferometry.Interferometer\n",
```
Comment (rocco8773) [inline docs/notebooks/diagnostics/line_integrated_diagnostic.ipynb:99]:
```suggestion
    "[Interferometer]: ../../api/plasmapy.diagnostics.path_integrated_diagnostic.Interferometer.rst#plasmapy.diagnostics.path_integrated_diagnostic.Interferometer\n",
```

This should fix the last error.

**Comment by pheuer** [comment]
Thanks for the help, documentation wizard @rocco8773 !

**Review by StanczakDominik — APPROVED** [review]
All right, this looks great to me and I could see it being merged - unless @rocco8773 wants to take another sweep, since this falls under diagnostics? (if no, just hit merge!)

Also, is it just me, or does this look like the Batman logo?

**Comment by pheuer** [comment]
Great! I don't see the batman logo even if I squint really really hard...I think you might need some sleep @StanczakDominik  ;)
